### PR TITLE
Issue 97: Support for Alternative Representations

### PR DESCRIPTION
**Pull request by arran4**
See my comment: https://github.com/arran4/golang-ical/issues/97# 

For details

Waiting review of original posters.

Also https://github.com/arran4/golang-ical/pull/73 is impacted sorry @ManoloTonto1 I will have to update my pr

Error additions and better matching of spec: https://icalendar.org/iCalendar-RFC-5545/3-2-1-alternate-text-representation.html

### CODE REVIEW

**Comment by masterElmar** [comment]
First Tests with Thunderbird as altrep in text/html works. Ill test some of the other altreps like chrome and plain.
But many changes and a lot of good refactoring and renaming ✅ 😊

**Comment by arran4** [comment]
image

Looks like I have some conflicts to resolve tomorrow too

**Comment by arran4** [comment]
That was a painful conflict resolution. Let me know how further testing goes please.

How long should I wait for?

**Comment by masterElmar** [comment]
> That was a painful conflict resolution for such a small difference. Let me know how further testing goes please.
> 
> How long should I wait for?

Thank you for all the effort 👍
Ill would test it the next 24h and then you could go on.
As far as i tested my Google and Thunderbird cases everything passes and after import of ical looks right.

**Comment by arran4** [comment]
Awesome. Thanks. I will look for a Saturday (Australian Eastern Standard time) merge.

**Comment by arran4** [comment]
Are you happy to wait for the other PRs, or do you want a semantic release for these changes this weekend? (The other PRs could take until the end of the year.)

**Comment by masterElmar** [comment]
> Are you happy to wait for the other PRs, or do you want a semantic release for these changes this weekend? (The other PRs could take until the end of the year.)

A release in the near future would be nice, because we use this library in our open source project for student schedules.
The new feature with altreps for diffrent clients is very helpful in our case.

**Comment by arran4** [comment]
Okay once tested, then a final test once released.

File: calendar.go [hunk 1]
```diff
@@ -404,25 +417,72 @@ func NewCalendarFor(service string) *Calendar {
...
+}
+
+func defaultSerializationOptions() *SerializationConfiguration {
+	serializeConfig := &SerializationConfiguration{
+		MaxLength:         75,
+		PropertyMaxLength: 75,
+		NewLine:           string(NewLine),
+	}
+	return serializeConfig
+}
```
Comment (masterElmar) [inline calendar.go:484]: Nice config options for the serializer 👍

File: components.go [hunk 2]
```diff
@@ -630,92 +651,100 @@ func (cal *Calendar) Todos() []*VTodo {
...
-func (c *VTodo) SetPriority(p int, params ...PropertyParameter) {
-	c.setPriority(p, params...)
+func (todo *VTodo) SetPriority(p int, params ...PropertyParameter) {
+	todo.setPriority(p, params...)
 }
 
-func (c *VTodo) SetResources(r string, params ...PropertyParameter) {
-	c.setResources(r, params...)
+func (todo *VTodo) SetResources(r string, params ...PropertyParameter) {
+	todo.setResources(r, params...)
```
Comment (masterElmar) [inline components.go:685]: Thats only refactoring, but improved readability is nice 👍

**Review by masterElmar — APPROVED** [review]

File: calendar.go (hunk 1 above)
Comment (arran4) [reply calendar.go:484]: Thanks

File: components.go (hunk 2 above)
Comment (arran4) [reply components.go:685]: It's work by @quite in PRs like: https://github.com/arran4/golang-ical/pull/105 He didn't miss any but there was perhaps a communication problem that caused a couple to get reverted. So I applied it here too.

**Comment by arran4** [comment]
Merged!

**Pull request merged**